File: Coding/python/schedule_service.py
"""Scheduled-task demo ported from scheduleService.java (Java package code.ds).

The Java version uses a ScheduledExecutorService and the SAME Runnable instance
is scheduled twice while its `delay` field is mutated to 15 before either task
fires -- so both runs observe 15 (shared mutable state). This port preserves
that behaviour with threading.Timer, but:
  * scales the delays down so the demo finishes quickly, and
  * joins the timers so the program terminates deterministically.
"""

import threading


class ScheduleService:

    # Java schedules in seconds; we shrink the unit to keep the demo fast.
    _SCALE = 0.01

    def __init__(self):
        self.delay = 0
        self._lock = threading.Lock()   # guard the shared `delay` field

    def run(self):
        with self._lock:
            print(self.delay)


if __name__ == "__main__":
    obj = ScheduleService()
    timers = []

    obj.delay = 10
    timers.append(threading.Timer(10 * ScheduleService._SCALE, obj.run))
    obj.delay = 15
    timers.append(threading.Timer(15 * ScheduleService._SCALE, obj.run))

    for t in timers:
        t.start()
    # Wait for all scheduled tasks to finish (analogous to awaiting shutdown).
    # Both fire after delay was set to 15, so the output is: 15 then 15.
    for t in timers:
        t.join()

File: Coding/golang/scheduleservice/scheduleservice.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// scheduleService mirrors the Java Runnable whose run() prints its delay field.
// The mutex makes concurrent reads/writes of delay safe (the Java original was
// unsynchronized but relied on a single shared object).
type scheduleService struct {
	mu    sync.Mutex
	delay int
}

func (s *scheduleService) setDelay(d int) {
	s.mu.Lock()
	s.delay = d
	s.mu.Unlock()
}

// run is the scheduled task body: print the current delay value.
func (s *scheduleService) run() {
	s.mu.Lock()
	d := s.delay
	s.mu.Unlock()
	fmt.Println(d)
}

func main() {
	obj := &scheduleService{}
	var wg sync.WaitGroup

	// schedule runs obj.run() once after the given delay, mirroring
	// ScheduledExecutorService.schedule. A WaitGroup replaces shutdown()+await
	// so the program terminates once queued tasks finish.
	schedule := func(after time.Duration) {
		wg.Add(1)
		time.AfterFunc(after, func() {
			defer wg.Done()
			obj.run()
		})
	}

	// Java scheduled the SAME object twice after mutating delay, so both runs
	// observe the final value (15). Durations are scaled to ms to keep the
	// demo fast and terminating.
	obj.setDelay(10)
	schedule(10 * time.Millisecond)
	obj.setDelay(15)
	schedule(15 * time.Millisecond)

	wg.Wait()
}
